src/chainstate/stacks/boot/pox_2_tests.rs

```diff
@@ -0,0 +1,322 @@
+};
+use types::proof::{ClarityMarfTrieId, TrieMerkleProof};
+
+use clarity_vm::clarity::Error as ClarityError;
+
+use super::test::*;
+
+const USTX_PER_HOLDER: u128 = 1_000_000;
+
+/// In this test case, two Stackers, Alice and Bob stack and interact with the
+///  PoX v1 contract and PoX v2 contract across the epoch transition.
+///
+/// Alice: stacks via PoX v1 for 4 cycles. The third of these cycles occurs after
+///        the PoX v1 -> v2 transition, and so Alice gets "early unlocked".
+///        After the early unlock, Alice re-stacks in PoX v2
+///        Alice tries to stack again via PoX v1, which is allowed by the contract,
+///        but forbidden by the VM (because PoX has transitioned to v2)
+/// Bob:   stacks via PoX v2 for 6 cycles. He attempted to stack via PoX v1 as well,
+///        but is forbidden because he has already placed an account lock via PoX v2.
+///
+#[test]
+fn test_simple_pox_lockup_transition_pox_2() {
+    let AUTO_UNLOCK_HT = 12;
+    let mut burnchain = Burnchain::default_unittest(0, &BurnchainHeaderHash::zero());
+    burnchain.pox_constants.reward_cycle_length = 5;
+    burnchain.pox_constants.prepare_length = 2;
+    burnchain.pox_constants.anchor_threshold = 1;
+    burnchain.pox_constants.v1_unlock_height = AUTO_UNLOCK_HT + 25;
+
+    let first_v2_cycle = burnchain
+        .block_height_to_reward_cycle(burnchain.pox_constants.v1_unlock_height as u64)
+        .unwrap()
+        + 1;
+
+    eprintln!("First v2 cycle = {}", first_v2_cycle);
+
+    let epochs = StacksEpoch::all(0, 25 + 10);
+
+    let (mut peer, mut keys) = instantiate_pox_peer_with_epoch(
+        &burnchain,
+        "test_simple_pox_lockup_transition_pox_2",
+        6002,
+        Some(epochs.clone()),
+    );
+
+    let num_blocks = 35;
+
+    let alice = keys.pop().unwrap();
+    let bob = keys.pop().unwrap();
+    let charlie = keys.pop().unwrap();
+
+    let mut alice_reward_cycle = 0;
+
+    for tenure_id in 0u32..num_blocks {
+        let microblock_privkey = StacksPrivateKey::new();
+        let microblock_pubkeyhash =
+            Hash160::from_node_public_key(&StacksPublicKey::from_private(&microblock_privkey));
+        let tip = SortitionDB::get_canonical_burn_chain_tip(&peer.sortdb.as_ref().unwrap().conn())
+            .unwrap();
+
+        let (burn_ops, stacks_block, microblocks) = peer.make_tenure(
+            |ref mut miner,
+             ref mut sortdb,
+             ref mut chainstate,
+             vrf_proof,
+             ref parent_opt,
+             ref parent_microblock_header_opt| {
+                let parent_tip = get_parent_tip(parent_opt, chainstate, sortdb);
+                let coinbase_tx = make_coinbase(miner, tenure_id as usize);
+
+                let mut block_txs = vec![coinbase_tx];
+
+                if tenure_id == 1 {
+                    // Alice locks up exactly 25% of the liquid STX supply, so this should succeed.
+                    // this locks for cycles 6, 7, 8, 9.
```

Comment:
Why is the first cycle -- i.e. at `tenure_id == 1` -- the 6th reward cycle?  Wouldn't it be the locked starting in the first reward cycle (if we start counting reward cycles from 0)?

Reply:
I think in the TestPeer, there's 25 bitcoin blocks before the first Stacks block.

Reply:
Ah, yes, that's right.